### PR TITLE
Improve Table Schema

### PR DESCRIPTION
**Pull request by beccadaniel**
## Why make this change?
The purpose of this change is to improve the current table schema. Doing this will help improve table performance and storage used.
An example of the inefficiencies is using `VARCHAR(max)` to store `custom_id` for the table created in `SQLServer_VectorStore`.
## What is this change?
- `custom_id` uses `varchar(1000)` instead of `varchar(max)`
- primary key `id` is non-clustered
- unique non-clustered index created on `custom_id`. This ensures that data with same `custom_id` cannot be entered into the table.
- `content_metadata` column uses `json` data type if the data type is available on the server, we determine this by checking if `type id` 244 is available in `sys.types`. In a scenario where `json` data type is not available, the data type of this column defaults to the value provided by sqlalchemy which is `NVARCHAR(max)`.

The primary key is made non-clustered because we use random values (uniqueidentifier).

If we were to have a clustered index, this would have been best suited for the custom_id column. However, there doesn't seem to be a need for this since the ultimate goal of this table is to do a semantic search on it. We don't need the entries to be sorted in any order, hence, the reason we have a non-clustered index on `custom_id`.

#### Sample generated create table query
```
CREATE TABLE  langchain_store_connection_test_v1 (
        id UNIQUEIDENTIFIER NOT NULL,
        custom_id VARCHAR(1000) NULL,
        content_metadata json NULL,
        content NVARCHAR(max) NOT NULL,
        embeddings VARBINARY(8000) NOT NULL CHECK (ISVECTOR(embeddings, 384) = 1),
        PRIMARY KEY NONCLUSTERED (id)
)

CREATE UNIQUE NONCLUSTERED INDEX idx_custom_id ON langchain_store_connection_test_v1 (custom_id)
```
image

## Tests
Additional test has been added to check that adding a row with pre-existing `custom_id` value in the table will error out.
image

image

### CODE REVIEW

File: libs/community/langchain_community/vectorstores/sqlserver.py [hunk 1]
```diff
@@ -194,6 +209,27 @@ class EmbeddingStore(DynamicBase):
 
         return EmbeddingStore
 
+    def _prepare_json_data_type(self) -> None:
```
Comment (anushakolan) [inline libs/community/langchain_community/vectorstores/sqlserver.py:263]: Do we have any advantage of using JSON value over VARCHAR?

File: libs/community/tests/unit_tests/vectorstores/test_sqlserver.py [hunk 2]
```diff
@@ -228,77 +228,41 @@ def test_that_add_text_fails_if_text_embedding_length_is_not_equal_to_embedding_
 
 def test_sqlserver_delete_text_by_id_valid_ids_provided(
     store: SQLServer_VectorStore,
+    texts: List[str],
```
Comment (anushakolan) [inline libs/community/tests/unit_tests/vectorstores/test_sqlserver.py:268]: Where are we declaring the deleted texts and metadatas?

File: libs/community/tests/unit_tests/vectorstores/test_sqlserver.py [hunk 3]
```diff
@@ -519,13 +471,34 @@ def test_that_case_sensitivity_does_not_affect_distance_strategy(
...
+
+
 # We need to mock this so that actual connection is not attempted
 # after mocking _provide_token.
 @mock.patch("sqlalchemy.dialects.mssql.dialect.initialize")
 @mock.patch(
     "langchain_community.vectorstores.sqlserver.SQLServer_VectorStore._provide_token"
 )
+@mock.patch(
+    "langchain_community.vectorstores.sqlserver.SQLServer_VectorStore._prepare_json_data_type"
```
Comment (anushakolan) [inline libs/community/tests/unit_tests/vectorstores/test_sqlserver.py:601]: I am trying to understand, why are we giving  prep_data_type: Mock, to only the entra authentication related tests and not others. Are these both related or am I missing something?

**Comment by anushakolan** [comment]
How did we identify that the below options improve the performance, do we have any other ways of improving also?
```
custom_id uses varchar(1000) instead of varchar(max)
primary key id is non-clustered
unique non-clustered index created on custom_id. This ensures that data with same custom_id cannot be entered into the table.
content_metadata column uses json data type if the data type is available on the server, we determine this by checking if type id 244 is available in sys.types. In a scenario where json data type is not available, the data type of this column defaults to the value provided by sqlalchemy which is NVARCHAR(max).
```

**Comment by beccadaniel** [comment]
> How did we identify that the below options improve the performance, do we have any other ways of improving also?
> 
> ```
> custom_id uses varchar(1000) instead of varchar(max)
> primary key id is non-clustered
> unique non-clustered index created on custom_id. This ensures that data with same custom_id cannot be entered into the table.
> content_metadata column uses json data type if the data type is available on the server, we determine this by checking if type id 244 is available in sys.types. In a scenario where json data type is not available, the data type of this column defaults to the value provided by sqlalchemy which is NVARCHAR(max).
> ```

They are more theoretical and the recommended ways of doing things.
- That VARCHAR(1000) will use less space than VARCHAR(max) is a fact.
- Since our primary key is a UUID, making it non-clustered improves the performance drawback. Non-clustered allows for efficient data retrieval, which we will be doing in the search context.

File: libs/community/langchain_community/vectorstores/sqlserver.py [hunk 4]
```diff
@@ -72,6 +76,7 @@ class DistanceStrategy(str, Enum):
 #
 EMBEDDING_LENGTH_CONSTRAINT = f"ISVECTOR(embeddings, :{EMBEDDING_LENGTH}) = 1"
 JSON_TO_ARRAY_QUERY = f"select JSON_ARRAY_TO_VECTOR (:{EMBEDDING_VALUES})"
+SERVER_JSON_CHECK_QUERY = "select name from sys.types where system_type_id = 244"
```
Comment (Aniruddh25) [inline libs/community/langchain_community/vectorstores/sqlserver.py:124]: Nice work in identifying the type_id here! #Closed

File: libs/community/langchain_community/vectorstores/sqlserver.py [hunk 5]
```diff
@@ -127,8 +132,7 @@ def __init__(
         self._bind: Union[Connection, Engine] = (
             connection if connection else self._create_engine()
         )
-        self._embedding_store = self._get_embedding_store(table_name, db_schema)
-        self._create_table_if_not_exists()
+        self._post_connection()
```
Comment (Aniruddh25) [inline libs/community/langchain_community/vectorstores/sqlserver.py]: what does it mean to post connection here? the function is actually creating the embedding store right?  #Closed

File: libs/community/langchain_community/vectorstores/sqlserver.py [hunk 6]
```diff
@@ -173,9 +182,15 @@ class EmbeddingStore(DynamicBase):
             """This is the base model for SQL vector store."""
 
             __tablename__ = name
-            __table_args__ = {"schema": schema}
+            __table_args__ = (
+                PrimaryKeyConstraint("id", mssql_clustered=False),
```
Comment (Aniruddh25) [inline libs/community/langchain_community/vectorstores/sqlserver.py:237]: if the primary key is also NON-clustered, there is no clustered index at all? What would the clustered index be if were to have it?

File: libs/community/langchain_community/vectorstores/sqlserver.py [hunk 7]
```diff
@@ -194,6 +209,27 @@ class EmbeddingStore(DynamicBase):
...
+        sqlalchemy. If it doesn't, this defaults to NVARCHAR(max) as specified
+        by sqlalchemy."""
+        try:
+            with Session(self._bind) as session:
+                result = session.scalar(text(SERVER_JSON_CHECK_QUERY))
+                session.close()
+
+                if result is not None:
+
+                    @compiles(JSON, "mssql")
```
Comment (Aniruddh25) [inline libs/community/langchain_community/vectorstores/sqlserver.py:275]: What does "mssql" do here?

Is this function changing the definition of the keyword JSON in sqlalchemy to mean the JSON type as opposed to NVARCHAR(max)? #Closed

File: libs/community/tests/unit_tests/vectorstores/test_sqlserver.py [hunk 8]
```diff
@@ -584,7 +561,11 @@ def test_that_given_a_connection_string_with_uid_and_pwd_entra_id_auth_is_not_us
...
+)
 def test_that_connection_string_with_trusted_connection_yes_does_not_use_entra_id_auth(
+    prep_data_type: Mock,
     provide_token: Mock,
     dialect_initialize: Mock,
 ) -> None:
```
Comment (Aniruddh25) [inline libs/community/tests/unit_tests/vectorstores/test_sqlserver.py:668]: General comment: I notice, the base branch of your fork has fallen way behind the community_package main, might be useful to bring it upto speed soon.

**Review by Aniruddh25 — APPROVED** [review]
LGTM, good idea on the check for JSON type

**Review by Aniruddh25 — CHANGES_REQUESTED** [review]
Can we do something similar to the JSON type for the vector type?

File: libs/community/langchain_community/vectorstores/sqlserver.py (hunk 1 above)
Comment (Aniruddh25) [reply libs/community/langchain_community/vectorstores/sqlserver.py:263]: yes, since the content is going to be of JSON type, the underlying storage will be MSJON that is optimized for storing data types like JSON..

File: libs/community/tests/unit_tests/vectorstores/test_sqlserver.py (hunk 2 above)
Comment (beccadaniel) [reply libs/community/tests/unit_tests/vectorstores/test_sqlserver.py:268]: texts and metadatas have been provided as fixtures (think of them as constants that we can re-use) at the beginning of the test. So, we don't have to redefine them inside the test, but instead provide it as a param the way we have here.

File: libs/community/tests/unit_tests/vectorstores/test_sqlserver.py (hunk 3 above)
Comment (beccadaniel) [reply libs/community/tests/unit_tests/vectorstores/test_sqlserver.py:601]: Yes, they are related.

For entra id authentication, recall that we need to call `_provide_token` to get auth token that then enables us to connect to the DB, however, we are now mocking the call to `_provide_token` to be able to check if it's called, this means we are not actually going through with the `_provide_token` call to get token and subsequently, every attempt to connect will fail.

`prep_json_data_type` attempts a connection to the DB in order to check for the data type, but we're already mocking an important part of that workflow and it's  bound to fail since we're not providing a valid token. This is why we mock `_prepare_json_data_type` too.

Comment (beccadaniel) [reply libs/community/tests/unit_tests/vectorstores/test_sqlserver.py:601]: Sure, these tests would be better suited for a test suite where the entirety of the test is based on Mocks. When we do create our mock tests, we can move this into those tests and modify as needed.

File: libs/community/langchain_community/vectorstores/sqlserver.py (hunk 7 above)
Comment (beccadaniel) [reply libs/community/langchain_community/vectorstores/sqlserver.py:275]: `mssql` indicates the dialect this function will have an effect on. So, the definition here simply means `JSON` keyword in `mssql` dialect should return what's defined in this function.

>Is this function changing the definition of the keyword JSON in sqlalchemy to mean the JSON type as opposed to NVARCHAR(max)?

Yes. By default, `NVARCHAR(max)` is what SQLAlchemy returns for `mssql` when we specify `JSON` as a column type.

File: libs/community/langchain_community/vectorstores/sqlserver.py (hunk 5 above)
Comment (beccadaniel) [reply libs/community/langchain_community/vectorstores/sqlserver.py]: Ah, this was me trying to see if I could reduce the number of mocks in the test to just one function. There is no specific need to have this function here. 

Yes, it calls a bunch of other functions.

Will revert back to original mode since this is unnecessary.

File: libs/community/langchain_community/vectorstores/sqlserver.py (hunk 6 above)
Comment (beccadaniel) [reply libs/community/langchain_community/vectorstores/sqlserver.py:237]: No, we do not have any clustered index.

The primary key makes sense as non-clustered because they are random values.

If we were to have a clustered index, this would have been best suited for the custom_id column. However, there doesn't seem to be a need for this since the ultimate goal of this table is to do a semantic search on it. We don't need the entries to be sorted in any order.

**Comment by beccadaniel** [comment]
> Can we do something similar to the JSON type for the vector type?

We can, but I don't yet have access to the Vector data type and can't tell what value it is yet, even with the traceflag enabled.

I could ask @yorek if he has any idea.

File: libs/community/langchain_community/vectorstores/sqlserver.py (hunk 6 above)
Comment (Aniruddh25) [reply libs/community/langchain_community/vectorstores/sqlserver.py:237]: can you please mention this as a comment in the PR description/in code why clustered index is not added

Comment (beccadaniel) [reply libs/community/langchain_community/vectorstores/sqlserver.py:237]: Sure

**Review by Aniruddh25 — APPROVED** [review]
LGTM! Track vector type changes as a separate PR.

**Review by anushakolan — APPROVED** [review]
LGTM!!

**Comment by beccadaniel** [comment]
> LGTM! Track vector type changes as a separate PR.

Done
